src/main/cpp/src/parse_uri.cu

```diff
@@ -23,279 +23,749 @@
- * @brief Parse protocol and copy from the input string column to the output char buffer.
- *
- * @tparam num_warps_per_threadblock Number of warps in a threadblock. This template argument must
- * match the launch configuration, i.e. the kernel must be launched with
- * `num_warps_per_threadblock * cudf::detail::warp_size` threads per threadblock.
- * @tparam char_block_size Number of characters which will be loaded into the shared memory at a
- * time.
+ * @brief Count the number of characters of each string after parsing the protocol.
  *
  * @param in_strings Input string column
- * @param in_validity Validity vector of output column
- * @param out_chars Character buffer for the output string column
- * @param out_offsets Offset value of each string associated with `out_chars`
+ * @param chunk Chunk of URI to return
+ * @param out_lengths Number of characters in each decode URL
+ * @param out_offsets Offsets to the start of the chunks
+ * @param out_validity Bitmask of validity data, updated in function
  */
-template <size_type num_warps_per_threadblock, size_type char_block_size>
-__global__ void parse_uri_to_protocol(column_device_view const in_strings,
-                                      bitmask_type* in_validity,
-                                      char* const out_chars,
-                                      size_type const* const out_offsets)
+__global__ void parse_uri_char_counter(column_device_view const in_strings,
+                                       URI_chunks chunk,
+                                       size_type* const out_lengths,
+                                       size_type* const out_offsets,
+                                       bitmask_type* out_validity)
 {
-  __shared__ char temporary_buffer[num_warps_per_threadblock][char_block_size];
-  __shared__ typename cub::WarpScan<int8_t>::TempStorage cub_storage[num_warps_per_threadblock];
-  __shared__ size_type out_idx[num_warps_per_threadblock];
-  __shared__ bool found_token[num_warps_per_threadblock];
-
-  auto const global_thread_id = cudf::detail::grid_1d::global_thread_id();
-  auto const global_warp_id   = static_cast<size_type>(global_thread_id / cudf::detail::warp_size);
-  auto const local_warp_id    = static_cast<size_type>(threadIdx.x / cudf::detail::warp_size);
-  auto const warp_lane        = static_cast<size_type>(threadIdx.x % cudf::detail::warp_size);
-  auto const nwarps     = static_cast<size_type>(gridDim.x * blockDim.x / cudf::detail::warp_size);
-  char* in_chars_shared = temporary_buffer[local_warp_id];
-
-  // Loop through strings, and assign each string to a warp
-  for (thread_index_type tidx = global_warp_id; tidx < in_strings.size(); tidx += nwarps) {
+  // thread per row
+  auto const tid      = threadIdx.x + blockIdx.x * blockDim.x;
+  auto const base_ptr = in_strings.child(strings_column_view::chars_column_index).data<char>();
+
+  for (thread_index_type tidx = tid; tidx < in_strings.size(); tidx += blockDim.x * gridDim.x) {
```

Comment:
I vaguely remember that there is some access function for `blockDim.x * gridDim.x`?